### DISCUSSION TITLE
Is tidy3d's adjoint method in time-domain or frequency-domain?

### DISCUSSION THREAD

**Discussion by Guizhixing** [flexcompute/tidy3d#1646] [Q&A] [answered]
Hi, all. 
I recently learned the lecture of Adjoint Method in learning center. The derivation of adjoint variable method i.e. AVM is starting from the solution of Maxwell’s equations in the frequency domain. In lecture, the objective function, the ajoint current source and gradient equation are expressed in frequency domain. 
Tidy3d is FDTD based program. As far as my knowledge of the FDTD based AVM，it considers the time marching scheme in FDTD as a dynamic system and uses time reversal to simulate the adjoint system. The objective function is restricted to a time-domain integral function. May I ask how tidy3d gets the gradient function by transfering frequency domain to time domain, or by any other means?

**Comment by tylerflex** [comment, marked as answer]
Hi @Guizhixing

You are correct, in the lectures we present the adjoint method in the frequency-domain. The reason is that it's much easier to reason about the adjoint method when the electric field `e` is written as the solution to a linear system `Ae=b`. This is a pretty common and natural way to write frequency-domain Maxwell's equations (although you can write time domain problems like this too).

In fact, you can extend the same adjoint analysis to the time domain, with a bit more complication. There is a paper we wrote a few years ago that derives everything for time and frequency domain here: https://arxiv.org/abs/1908.10507. The basic idea is the same, except technically the adjoint FDTD is run backwards in time and also the adjoint source is time-dependent, among other details. The time domain adjoint often also suffers from high memory requirements due to the need to store the entire time series info.

In tidy3d's `adjoint` plugin, we try to take the best of both worlds approach: We run FDTD in the time-domain, but use the `td.FieldMonitor` to store the electromagnetic fields in the frequency-domain at the frequencies of interest in the objective function. These fields are computed by performing a running Fourier transform of the time series, which is memory efficient.  Then, we use the frequency-domain adjoint formalism to compute the gradients.

Hopefully that makes sense but let me know if you need clarification or have other questions!

**Reply by Guizhixing** [reply]
I see how the Tidy3D works, thank you very much for your replies.

**Comment by Nikolaos-Matthaiakakis** [comment]
Hello, I was wondering what would be the trade-off in terms of performance when considering broadband goals (10+ frequency points). Would the time domain adjoint method still be a more demanding method when compared to frequency domain? Is there a plan to also implement a time domain adjoint optimizer in the future? Thank you

**Reply by Guizhixing** [reply]
Tidy3D is fully capable of broadband goal optimization. As long as the targeted bandwidth does not violate fundamental physical limits (such as the Chu-Harrington limit for antennas), you can achieve excellent performance across the spectrum by assigning appropriate weights to different frequency points.

Also, I think it's much easier to compute gradients from FDTD results when optimization objective are defined in the frequency domain. I actually wrote my own code for antenna pattern optimization using this hybrid domain AVM. If we tried to do far-field pattern optimization using a full time-domain AVM, the sensitivity formulas would just be too complex.

**Reply by Nikolaos-Matthaiakakis** [reply]
Thanks for the reply. It is just that sometimes when one studies the phase or group delay there can be very sharp changes that require a lot of frequency points, thus resulting in multiple adjoint simulation runs. I saw this interesting paper https://onlinelibrary.wiley.com/doi/10.1002/lpor.202401819 that instead uses time domain optimization.
